RZ_A1LU: Fix TRNG function

TomoYamanaka opened the pull request:
Related to the review of #5857, I fixed the TRNG function for GR-LYCHEE.
- I modified to zeroize "recv_data" before the function return.
- I added the processing that check the return value of I2C.read function. If return value is error, "output"  is zeroized before  function return.

## Related PRs
https://github.com/ARMmbed/mbed-os/pull/5857


## Request
@cmonr 
Could you revert #5967 when this PR passed?

hanno-arm commented on targets/TARGET_RENESAS/TARGET_RZ_A1XX/TARGET_GR_LYCHEE/trng_api_esp32.cpp:
```diff
@@ -80,6 +85,10 @@ extern "C" int trng_get_bytes_esp32(uint8_t *output, size_t length, size_t *outp
+    if (idx == 0) {
+        memset(output, 0, length);
+    }
+    memset(recv_data, 0, sizeof(recv_data));
```
This might be optimized away by the compiler, please consider using something similar to `mbedtls_zeroize`.

hanno-arm commented on targets/TARGET_RENESAS/TARGET_RZ_A1XX/TARGET_GR_LYCHEE/trng_api_esp32.cpp:
```diff
@@ -65,13 +65,18 @@ extern "C" int trng_get_bytes_esp32(uint8_t *output, size_t length, size_t *outp
         ret = mI2c.write(ESP32_I2C_ADDR, send_data, 1);
         if (ret == 0) {
-            mI2c.read(ESP32_I2C_ADDR, recv_data, sizeof(recv_data));
+            ret = mI2c.read(ESP32_I2C_ADDR, recv_data, sizeof(recv_data));
```
Why are we maintaining an error counter for the `write` calls, but none for `read`?

cmonr commented:
@TomoYamanaka @hanno-arm Thank you for being so prompt with this PR!

@TomoYamanaka I'm holding off on merging in https://github.com/ARMmbed/mbed-os/pull/5967 for now, and won't merge unless this PR takes too long to get into 5.7.5 (two-ish weeks from now).

TomoYamanaka commented:
@cmonr 
Thank you for your cooperation. I will address the above comments asap.

TomoYamanaka commented:
@hanno-arm
I rebased the commit due to address your two comments.

1.
I changed the way of zeroization of buffer from `memset` to `mbedtls_zeroise` by referring to another vendor.

2.
`err_cnt` is used as a retry counter. There is a time lag in the period from ESP32 reset to start working, error may occur when `Write` is called.
Thus, I added a retry counter due to address this concern. there is not this counter for `Read` since it is called after `Write`.
I renamed `err_cnt` because this name does not express behavior.

hanno-arm commented:
@0xc0170 Could you also request a review from @mazimkhan please? Thanks!

mazimkhan commented on targets/TARGET_RENESAS/TARGET_RZ_A1XX/TARGET_GR_LYCHEE/trng_api_esp32.cpp:
```diff
@@ -59,19 +64,24 @@ extern "C" int trng_get_bytes_esp32(uint8_t *output, size_t length, size_t *outp
+            ret = mI2c.read(ESP32_I2C_ADDR, recv_data, sizeof(recv_data));
+            if (ret != 0) {
+                idx = 0;
+                break;
```
No retry in case of read error. How is it different then a write error?

hanno-arm replied:
@mazimkhan There has already been some discussion on this. See above.

mazimkhan commented on targets/TARGET_RENESAS/TARGET_RZ_A1XX/TARGET_GR_LYCHEE/trng_api_esp32.cpp:
```diff
@@ -59,19 +64,24 @@ extern "C" int trng_get_bytes_esp32(uint8_t *output, size_t length, size_t *outp
 
     while (idx < length) {
         send_data[0] = 0;
         ret = mI2c.write(ESP32_I2C_ADDR, send_data, 1);
```
`1` -> `sizeof( send_data )`

mazimkhan commented on targets/TARGET_RENESAS/TARGET_RZ_A1XX/TARGET_GR_LYCHEE/trng_api_esp32.cpp:
```diff
@@ -59,19 +64,24 @@ extern "C" int trng_get_bytes_esp32(uint8_t *output, size_t length, size_t *outp
         } else {
-            err_cnt++;
-            if (err_cnt >= 20) {
+            retry_cnt++;
```
Mixing retry logic within this loop is not clean and it does not cover read error (if that is also required to be covered). I suggest having a supervising retry loop with condition ```for ( retry = 0; retry < 20 && idx < length; retry++ )```.

mazimkhan commented on targets/TARGET_RENESAS/TARGET_RZ_A1XX/TARGET_GR_LYCHEE/trng_api_esp32.cpp:
```diff
@@ -80,6 +90,10 @@ extern "C" int trng_get_bytes_esp32(uint8_t *output, size_t length, size_t *outp
     if (output_length != NULL) {
```
If ```output_length``` is a required out parameter then ```output_length == NULL``` is an invalid input. In that case it should be tested in the beginning of the function.

TomoYamanaka commented:
@mazimkhan 
I rebased the commit due to address your comments.

<br>

> ret = mI2c.write(ESP32_I2C_ADDR, send_data, 1); 
> 
> 1 -> sizeof( send_data )

I modified from `1` to `sizeof( send_data )` .


> Mixing retry logic within this loop is not clean and it does not cover read error (if that is also required to be covered). I suggest having a supervising retry loop with condition for ( retry = 0; retry < 20 && idx < length; retry++ ).

This retry loop is not required to cover read error, I think that read error does not occur usually in this situation. However, just in case, I accepted your suggestion and modified this retry logic.


> If output_length is a required out parameter then output_length == NULL is an invalid input. In that case it should be tested in the beginning of the function.

I added the NULL check process at the beginning of the function.

0xc0170 commented:
/morph build

0xc0170 commented:
/morph uvisor-test

0xc0170 commented:
/morph uvisor-test

cmonr commented:
/morph uvisor-test

cmonr commented:
/morph uvisor-test

cmonr commented:
Waiting on https://github.com/ARMmbed/mbed-os/pull/6026 to be merged to retest.
PR removes failing device in failing test.

cmonr commented:
/morph uvisor-test

cmonr commented:
/morph uvisor-test
